### PR TITLE
Feature/33962 custom subsys

### PR DESCRIPTION
**Pull request by gregshue**

### CODE REVIEW

File: include/custom_subsys/custom_subsys.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+#include <stdint.h>
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @defgroup isr_apis Interrupt Service Routine APIs
+ * @ingroup kernel_apis
```
Comment (gregshue) [inline include/custom_subsys/custom_subsys.h:22]: What are the preferred groups to put this in?

File: samples/subsys/app_empty_main/README.rst [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+********************
+
+This application can be built and executed on QEMU Cortex M0 as follows:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/app_empty_main
+   :host-os: unix
+   :board: qemu_cortex_m0
+   :goals: run
+   :compact:
```
Comment (gregshue) [inline samples/subsys/app_empty_main/README.rst:27]: These commands need to be verified.  How?

File: samples/subsys/app_empty_main/sample.yaml [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+  description: App Empty Main, sample made entirely from
+    subsystems and drivers, and no console output
+  name: App Empty Main
+common:
+    tags: introduction
+    integration_platforms:
+      - native_posix
+tests:
+  sample.basic.app_empty_main:
+    tags: introduction
```
Comment (gregshue) [inline samples/subsys/app_empty_main/sample.yaml:11]: Is there a better tag than `introduction`?

File: subsys/custom_subsys/custom_subsys.c [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+static int init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	LOG_INF("Hello from " STRINGIFY(MODULE_NAME));
+
+	return 0;
+}
+
+SYS_INIT(init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
```
Comment (gregshue) [inline subsys/custom_subsys/custom_subsys.c:28]: How should the init priority be set?

File: zephyr/module.yml [hunk 5]
```diff
@@ -1,6 +1,8 @@
 # Copyright (c) 2021 Nordic Semiconductor ASA
 # SPDX-License-Identifier: Apache-2.0
 
+name: example-application
```
Comment (gregshue) [inline zephyr/module.yml:4]: Please suggest text describing this setting, covering:
How is this setting used (e.g., used to derive directory names?, injected into documentation text?)
What restrictions are there no the value (e.g., no spaces?, all lower-case?, only alphanumeric, underscore, hyphen?)

File: zephyr/module.yml [hunk 6]
```diff
@@ -18,3 +20,13 @@ build:
...
     dts_root: .
+
+tests:
+  # Path(s) to the directory tree(s) containing `testcase.yaml` files.
+  # This is for use by `twister` to discover testcases to run.
+  - ./tests
+
+samples:
+  # Path(s) to the directory tree(s) containing `sample.yaml` files.
+  # This is for use by <TBD> to discover samples to <TBD>
```
Comment (gregshue) [inline zephyr/module.yml:31]: Please suggest text describing this setting.

**Review by mbolivar-nordic — COMMENTED** [review]
I'm really not trying to get into a semantic quagmire here. but I'm not sure what you mean by "subsystem".

This doesn't seem to be adding a new driver API, which I think is a good thing, because last time I checked we couldn't add syscall support for any APIs defined outside the zephyr repository. That may be stale knowledge, though.

So maybe you mean a subsystem like Bluetooth is a subsystem? I'm kind of confused about what value this adds apart from a custom library. I'm also confused about why there is an application inside of a subsystem. I don't think anything in zephyr/subsys is an application (other than the ztest one, I guess) and I'm not sure what purpose this serves.

**Comment by gregshue** [comment]
@mbolivar-nordic In this case, "subsystem" is anything within `subsys/`.  In general this might be any logic that may be integrated into multiple projects through CONFIG_* settings.  This adds two things of value:

1. it provides an example of a header file within the specified module-level `include/` directory to demonstrate visibility,
2. it provides an example of how the specified module-level `subsys/` directory is added into the build.

> I'm also confused about why there is an application inside of a subsystem. I don't think anything in zephyr/subsys is an application (other than the ztest one, I guess) and I'm not sure what purpose this serves.

What is an application, other than a stand-alone service and/or agent?  We already have those in IO stacks (e.g., ping, echo, DHCP servers), as well as natively (e.g., embedded-web-servers, UIs, periodically polling the Internet for FW updates), consoles, shell, USB devices (e.g. DFU, network interface).  In my experience it is counterproductive to assume any functionality will only ever be unique to one product, so I put all logic in a place it can be pulled into multiple executables.

How do you distinguish between an "application" and a "subsystem", and what led you to that preference?

**Comment by mbolivar-nordic** [comment]
Let me put this another way. Why are you creating a subsys/ directory and putting an application in there? That is contrary to how things are normally done in zephyr.

**Comment by gregshue** [comment]
@mbolivar-nordic

> Why are you creating a subsys/ directory and putting an application in there? 
 ... for the same reason that the Shell, console, DFU, IPC, and mcumgr are in the `subsys/` directory.

I have previously worked on multi-function printers (print/copy/scan/fax/email/external files storage/etc) that ran on a 32b ARM-based SOC running an RTOS.  It could simultaneously process 5+ different jobs involving independent sets of HW resources, not to mention the simultaneous status/controls across every communication interface.  It became painfully clear that in many embedded devices there is not one "application" but many found across multiple layers of the ISO model.  

Until we can agree on a clear distinction between an "application" and a "subsystem", there are no limitations on what can go in  `subsys/`.  Please try to answer my earlier question about how you distinguish between an "application" and a "subsystem".

**Comment by galak** [comment]
@gregshue you may need to rebase now that we committed change to remove the github workflow cache.  Should hopefully pass CI if you do.

**Comment by gregshue** [comment]
The CI for `example-application` has been fixed, so please re-review this PR.

File: subsys/custom_subsys/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,11 @@
+# SPDX-License-Identifier: Apache-2.0
+
+zephyr_interface_library_named(custom_subsys)
```
Comment (carlescufi) [inline subsys/custom_subsys/CMakeLists.txt:3]: Let us use the same approach that we used in the lib:

```
zephyr_library()
zephyr_library_sources(custom_subsys.c)
zephyr_include_directories(${ZEPHYR_CURRENT_MODULE_DIR}/include/custom_subsys)
```

File: subsys/custom_subsys/custom_subsys.c (hunk 4 above)
Comment (carlescufi) [reply subsys/custom_subsys/custom_subsys.c:28]: `APPLICATION` is fine for now.

File: subsys/custom_subsys/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+zephyr_library()
+zephyr_library_sources(custom_subsys.c)
+
+zephyr_library_link_libraries(custom_subsys)
+
+add_definitions(-DMODULE_NAME="${module_name}")
```
Comment (carlescufi) [inline subsys/custom_subsys/CMakeLists.txt:10]: Use `add_compile_definitions()` instead, since `add_definitions()` is superseeded by it.

**Review by carlescufi — CHANGES_REQUESTED** [review]

File: subsys/custom_subsys/CMakeLists.txt (hunk 7 above)
Comment (gregshue) [reply subsys/custom_subsys/CMakeLists.txt:3]: Accepted and manually integrated.

File: subsys/custom_subsys/CMakeLists.txt (hunk 8 above)
Comment (gregshue) [reply subsys/custom_subsys/CMakeLists.txt]: Accepted, and manually rolled in.

**Review by gmarull — CHANGES_REQUESTED** [review]
-1 on this one. A subsys is nothing else than a library, and we already merged a sample library.

**Comment by gregshue** [comment]
By the same logic an app and a test are nothing else than a sample, and a driver is nothing else than a library.  Yet the reason each of those exist here is because they are different sub-trees in the Zephyr directory structure. One of the primary purposes of this example is to provide a template for downstream users to fill in for their OOT modules.  In particular, to show how to properly fill out `module.yml` for each type of addition.

**Comment by gmarull** [comment]
> By the same logic an app and a test are nothing else than a sample, and a driver is nothing else than a library. Yet the reason each of those exist here is because they are different sub-trees in the Zephyr directory structure. One of the primary purposes of this example is to provide a template for downstream users to fill in for their OOT modules. In particular, to show how to properly fill out `module.yml` for each type of addition.

this is being too opinionated IMO. Applications can place libraries wherever they want: lib, subsys, mylib, etc. We just need to show how you can create a library. Drivers are also a library, but when extending upstream driver classes, you need to respect the same folder layout to make things work smoothly.

**Comment by gregshue** [comment]
> this is being too opinionated IMO.

You are certainly entitled to your opinion, as I am mine. I am offering this as something I as a downstream user was looking for guidance on and expected to find in this example. 

> Applications can place libraries wherever they want ... We just need to show how you can create a library.

Just showing how you can create a library is not showing best practice. It is simply showing possibilities. I am looking here for best practice on how to set up and use the subsys/driver/lib/etc. folders - so that I have the highest chance of being aligned with whatever capabilities and tools come along from Zephyr.

**Comment by gmarull** [comment]
stale, closing

**Pull request closed without merging**